### DISCUSSION TITLE
How to extract model strain energy and kinetic energy

### DISCUSSION THREAD

**Discussion by ThresholDDD** [sofa-framework/sofa#2702] [Develop in C++] [answered]
Hi,
I want to get the kinetic energy and strain energy of my model during the linear FEM simulation (SimpleFEM  component).  I am thinking in two ways. One is to get mass matrix and stiffness matrix because I know how to get the velocities and displacement. So the question is how to get globall mass matrix and stiffness matrix? or in another approach, directly get two energies. Do Sofa has access to energy directly?

**Comment by alxbilger** [comment, marked as answer]
Hi,

regarding accessing the matrices: it is now possible to export the global system matrix (https://github.com/sofa-framework/sofa/blob/master/applications/plugins/SofaMatrix/examples/GlobalSystemMatrixExporter.scn). Both mass and stiffness matrices are merged into a single matrix. If you really need to access both matrices distinctly, you'll have to wait. This is in the [roadmap](https://www.sofa-framework.org/consortium/roadmap/) and planned for next release.

Regarding accessing the energy: masses have the function `getKineticEnergy` and force fields have the function `getPotentialEnergy`. I don't guarantee that all force fields and masses implement correctly those two functions. You can display a graph of the evolution of each energy in a Node:
![image](https://user-images.githubusercontent.com/10572752/154941077-4f11bd7f-006b-417b-95a8-9d5dcf4d70f9.png)
In runSofa, just double click on any Node in your scene graph and go into the "Energy Stats" tab (`SOFAGUIQT_HAVE_QT5_CHARTS` must be enabled).
If you want to export them, I think you must use the component `ExtraMonitor` (https://www.sofa-framework.org/community/doc/using-sofa/inspect-components/).

**Reply by AndyRay1998** [reply]
Hi,

I am using SOFA v22 and I dont seem to have an "Energy Stats" tab whichever scene I try to run. Do you know any potential reasons for that? I also tried to add an additional variable in c++ source code to retreive the energy, but the variable is not updated at all.

Thanks a lot!
CY

**Comment by aleblanc30** [comment]
HI @alxbilger, sorry to revive an old thread. Googling brought me here so I thought others with my question would end up here as well. Do you know if there is any way to access kinetic and potential energy from a SofaPython scene ? If none is available, I guess one will have to go the route described in this other thread : https://github.com/sofa-framework/sofa/discussions/2601

**Reply by alxbilger** [reply]
No I don't see any way to access the energies in Python. What I would do:

1. Create a binding in Node (https://github.com/sofa-framework/SofaPython3/blob/master/bindings/Sofa/src/SofaPython3/Sofa/Core/Binding_Node.cpp) to a free function `computeEnergy`
2. In this free function, I would do:
```cpp
MechanicalComputeEnergyVisitor energyVisitor(&mparams);
executeVisitor(&energyVisitor);
kineticEnergy = energyVisitor.getKineticEnergy();
potentialEnergy = energyVisitor.getPotentialEnergy();
```
3. Create a Pull Request 😃

**Comment by aleblanc30** [comment]
Thanks ! I'll look into it.
________________________________
